Add Support for View of vector of Ptr

wddgit opened the pull request:
Add support for View of vector of Ptr. This completes implementation of this new feature. There is a test of its usage in the class ViewAnalyzer in function testStdVectorPtr which is run by ViewTest_cfg.py in the package FWCore/Integration. This does change interfaces. Code in the release was fixed to compile and so that runTheMatrix will pass. But runtime problems could exist in code not run in runThe Matrix and compile time problems for code not in the release. For example, it is no longer possible to get the container product from a Ref which implies many things. For example on cannot construct a RefProd from a Ref. One change in particular that might cause problems. If one inserts into an AssociationMap constructed with the no argument constructor an exception will be thrown at runtime, There is known code like this in the release, although none is called in runTheMatrix tests.

Dr15Jones committed:
No longer construct a RefToBaseProd from a View

Views no longer give access to ProductID or DataProductGetter
so one can no longer construct a RefToBaseProd from a View.

Dr15Jones committed:
View now internally uses an std::vector<edm::Ptr<>>

In order to support reading an std::vector<edm::Ptr<>> into a
View it was necessary to remove the use of edm::PtrVector and
edm::RefToBaseVector from View.
This also required the ability to create edm::RefToBase<> objects
out of edm::Ptr instances.

Dr15Jones committed:
Move HolderToVectorTrait specializations to own files

To avoid unnecessary clutter in header files and to allow faster compilation for the case where the specialization isn't needed, the specializations of HolderToVectorTrait were moved to their own files.
The use of the other header's include guards to guarantee we only include the code when the other two files are included is unorthodox but works in this case since all files live in the same package.

Dr15Jones committed:
Add missing headers

Dr15Jones committed:
Switch to FillViewHelperVector for View

In order to support reading a std::vector<edm::Ptr<>> into a View, we need to record additional information.
In addition, the code no longer needs to use RefHolder based classes.

Dr15Jones committed:
Allow std::vector<edm::Ptr<T>> to be read by a edm::View<T>

This change sets up all the code needed to switch on reading an std::vector<edm::Ptr<T>> by an edm::View<T>.

Dr15Jones committed:
Add test of reading std::vector<edm::Ptr<int>> via edm::View<int>

In addition to the test addition, the code was also made to conform to the new threaded framework.

Dr15Jones committed:
Added constructor that takes an std::unique_ptr

Dr15Jones committed:
Removed inappropriate use of castTo and fixed getByToken problem

-The edm:RefToBase<>::castTo was being used as a complicated and slow
way of doing a dynamic cast of the element of the container.
-The optional call to getByToken for std::vector<Trajectory> was using
the same GetByTokenT<> as for the Tracks. This can not work since they
are different types. Changed to use a separate container of tokens for
the trajectories.
-Removed unnecessary copies of containers from the Event.
-Used a reference to the Track directly rather than making a unnecessary
TrackRef.

Dr15Jones committed:
Removed inappropriate use of castTo

The code was using edm::RefToBase<>::castTo as a complicated and
slow method of doing a dynamic_cast of the element of the container.
Doing the dynamic_cast directly is cleaner and faster as well as
avoids future complications with castTo.

Dr15Jones committed:
Allow Ptr and RefToBase to be used for lookup in AssociationVector

By allowing Ptr and RefToBase to be used to lookup items with an
AssociationVector we can avoid an unnecessary creation of a Ref.

Dr15Jones committed:
Allow Ptr and RefToBase to be used for lookup in AssociationVector

By allowing Ptr and RefToBase to be used to lookup items with an
AssociationVector we can avoid an unnecessary creation of a Ref.

Dr15Jones committed:
Allow perfect forwarding of KeyVal for AssociationMap

Use C++11 perfect forwarding to handle constructing of key and value
components from allowed constructor arguments. This makes creating an
AssociationMap easier.

Dr15Jones committed:
Allow perfect forwarding of KeyVal for AssociationMap

Use C++11 perfect forwarding to handle constructing of key and value
components from allowed constructor arguments. This makes creating an
AssociationMap easier.

Dr15Jones committed:
Require that AssociationMap be initialized before doing an insert

It is now a requirement that AssociationMap be initialized with information
about which container(s) it will use.

Dr15Jones committed:
Require that AssociationMap be initialized before doing an insert

It is now a requirement that AssociationMap be initialized with information
about which container(s) it will use.

Dr15Jones committed:
AssociationMap must be initialized and can use Ptr and RefToBase for lookup

Changed AssocationMap so that it is possible to lookup items use a
Ptr or a RefToBase. This avoids requiring users to create a edm::Ref
just for lookup.
Added a constructor to AssociationMap to allow one to associate which
collection(s) the container will use for association. This constructor
must be used since it is no longer possible to have those collections
be set on the first insert to the container.

Dr15Jones committed:
AssociationMap must be initialized and can use Ptr and RefToBase for lookup

Changed AssocationMap so that it is possible to lookup items use a
Ptr or a RefToBase. This avoids requiring users to create a edm::Ref
just for lookup.
Added a constructor to AssociationMap to allow one to associate which
collection(s) the container will use for association. This constructor
must be used since it is no longer possible to have those collections
be set on the first insert to the container.

Dr15Jones committed:
No longer create a edm::RefProd from a edm::Ref

The design which allowed a edm::RefProd to be constructed from an
edm::Ref caused problems for future changes (edm::View and slimming).
Given this was equivalent to taking an arbitrary pointer and then
obtaining which std::vector<> the data is stored in, we determined
this was not an absolute requirement.

Dr15Jones committed:
No longer create a edm::RefProd from a edm::Ref

The design which allowed a edm::RefProd to be constructed from an
edm::Ref caused problems for future changes (edm::View and slimming).
Given this was equivalent to taking an arbitrary pointer and then
obtaining which std::vector<> the data is stored in, we determined
this was not an absolute requirement.

Dr15Jones committed:
No longer create a edm::RefToBaseProd from a edm::Ref or RefToBase

The design which allowed a RefToBaseProd to be constructed from an
edm::Ref or RefToBase caused problems for future changes (edm::View
and slimming). Given this was equivalent to taking an arbitrary pointer
and then obtaining which std::vector<> the data is stored in, we
determined this was not an absolute requirement.

Dr15Jones committed:
No longer create a edm::RefToBaseProd from a edm::Ref or RefToBase

The design which allowed a RefToBaseProd to be constructed from an
edm::Ref or RefToBase caused problems for future changes (edm::View
and slimming). Given this was equivalent to taking an arbitrary pointer
and then obtaining which std::vector<> the data is stored in, we
determined this was not an absolute requirement.

wddgit committed:
Ref cache points at element not container

Makes the pointer cached in a Ref point to the
element in the container instead of pointing
to the container. Removes the product function
from classes related to Ref, RefToBase, Ptr, and
vectors of them. Remove constructor of Ref from
RefVector. Add Ref constructor that takes ProductID,
pointer to element, and a key. Remove RefProd
constructor from Ref and RefProd constructor from
RefVector. Remove  RefToBaseProd constructors from
Ref, RefToBase, and View. Make RefToBaseProd from
RefProd constructor explicit.

This is all motivated by changes to the View class
that indirectly require this to work properly.

Many changes to make code that depends on these changes
continue work as before.

Additions to make it easier to build AssociationMap
with a one argument constructor.

A previous commit required the AssociationMap be
initialized with RefProd before calling insert. This
commit fixes some but not all of the cases where
this caused failures.

Note this is work in progress. If you try to run with
this commit failures will occur.

wddgit committed:
Merge branch 'avoidUsingRefInConstructor' of https://github.com/Dr15Jones/cmssw into removeProductFunction4

wddgit committed:
Fixes related to AssociationMap in RecoMuon

wddgit committed:
Merge branch 'CMSSW_7_4_X' of https://github.com/cms-sw/cmssw into removeProductFunction4

wddgit committed:
Finish implementing RefVector member pointers

wddgit committed:
Another change related to AssociationMap migration

wddgit committed:
Merge branch 'CMSSW_7_4_X' of https://github.com/cms-sw/cmssw into removeProductFunction4

wddgit committed:
RefToBaseProd migration in MasterCollection

MasterCollection now needs the Event to make the
RefToBaseProd in MasterCollection.

wddgit committed:
Extend unit tests

Extend unit tests for case of reading a RefVector
into a View, then converting to RefToBaseProd.

wddgit committed:
Migrate TrackExtraBase for Ref changes

wddgit committed:
Protect Ref when constructed invalid

wddgit committed:
AssociationMap migration

Need to initialize AssociationMap products
in the constructor now.

wddgit committed:
Adjust PFCandidate for redesign of Ref

PFCandidate caches the guts of Refs and it
matters to the cache that the pointer is
changing from pointing at the container to
pointing at an element in the container.
Modify it to work with the new Ref implementation.

wddgit committed:
Migrate code to new Association interface

Migrate code to new interfaces for AssociationMap
and AssociationVector. This is one in a series of
commits related to this migration and part of this
commit fixes some of the migration done in previous
commits.

wddgit committed:
Merge commit '60f95c1' into removeProductFunction4

Conflicts:
	DataFormats/Common/interface/FillView.h
	DataFormats/Common/interface/OwnVector.h
	DataFormats/Common/interface/RefProd.h
	DataFormats/Common/interface/RefToBaseProd.h
	DataFormats/Common/interface/RefVector.h
	DataFormats/Common/interface/RefVectorHolderBase.h
	DataFormats/Common/test/ref_t.cppunit.cc
	FWCore/Framework/test/stubs/ToyRefProducers.cc
	FWCore/Integration/test/ViewTest_cfg.py

wddgit committed:
Finish resolving merge conflicts

wddgit committed:
Add missing #include

wddgit committed:
Add productGetter accessor

Add productGetter accessor so client code
using fillView_ does not need to include
the EventPrincipal header to convert the
pointer from EventPrincipal* to
EDProductGetter*.

wddgit committed:
Merge branch 'removeProductFunction4' of github.com:wddgit/cmssw into viewWithVectorOfPtr

Conflicts:
	DataFormats/Common/interface/Wrapper.h

wddgit committed:
Fix unit tests

wddgit committed:
Extend RefToBase::castTo function

Extend the castTo function to be able to handle
casting in more ways, in particular we need to
be able to cast from a contained Ptr to a Ref.

wddgit committed:
Migrate to modified View interface

wddgit committed:
Remove unnecessary GCCXML protection

wddgit committed:
Add necessary #includes and forward declarations

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr

wddgit committed:
Delete unused function

wddgit committed:
Delete unused function

wddgit committed:
Delete unused code

wddgit committed:
Migrate to new AssociationMap interface

wddgit committed:
New AssociationMap constructor

Add new AssociationMap constructor that takes
an EDProductGetter. This allows initializing
the RefProd(s) or RefToBaseProd(s) on the first
insert.

wddgit committed:
Merge branch 'viewWithVectorOfPtr' of github.com:wddgit/cmssw into viewWithVectorOfPtr2

wddgit committed:
Add missing dictionaries

Changes to Core Ref related classes add the need
for dictionaries to RefHolder<Ptr<X>> for several
types X.

wddgit committed:
Migrate clients to new AssociationMap interface

wddgit committed:
Progress towards AssociationMap unit tests

wddgit committed:
Add missing dictionary

wddgit committed:
increment version of RefVectorBase

wddgit committed:
Migrate to new AssociationMap interface

wddgit committed:
Add to AssociationMap unit test

wddgit committed:
Add to AssociationMap unit test

wddgit committed:
Add to unit tests

wddgit committed:
Merge branch 'viewWithVectorOfPtr2' of github.com:wddgit/cmssw into viewWithVectorOfPtr3

wddgit committed:
Add to unit tests and minor cleanup

wddgit committed:
Minor cleanup of headers and spacing

wddgit committed:
Change argument signature of new Ptr constructor

wddgit committed:
Add to AssociationVector unit test

wddgit committed:
Delete unused function makeVectorBaseHolder

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr3

wddgit committed:
Only one Ptr constructor taking a pointer to item

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr3

wddgit committed:
Add to unit tests

wddgit committed:
Fix to be consistent with modified Ptr constructor

wddgit committed:
Cleanup header includes, forward declarations, white space

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr3

wddgit committed:
Migrate PackedCandidate for modified Ref interface

product function is removed from the edm::Ref class.
Modify PackedCandidate and things that depend on it
for this.

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr3

Conflicts:
	DataFormats/Common/interface/RefVector.h
	DataFormats/Common/interface/RefVectorHolderBase.h
	DataFormats/TestObjects/src/classes.h
	DataFormats/TestObjects/src/classes_def.xml

wddgit committed:
Fix merge with ROOT 6 changes

wddgit committed:
Handle empty case more cleanly

wddgit committed:
Update version and checksum for PackedCandidates

wddgit committed:
Fix duplicate dictionary issue created in merge

wddgit committed:
Fix version number

Increment version numbers to 13 for RefVector and
RefVectorBase for ROOT 6. Leave version 12 for
ROOT 5 branch.

wddgit committed:
Add ROOT 5 and ROOT 6 checksums for PackedCandidate

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr4

cmsbuild commented:
A new Pull Request was created by @wddgit (W. David Dagenhart) for CMSSW_7_5_X.

View with vector of ptr4

It involves the following packages:

AnalysisAlgos/TrackInfoProducer
Calibration/Tools
CommonTools/RecoAlgos
CommonTools/RecoUtils
CommonTools/UtilAlgos
DataFormats/CaloRecHit
DataFormats/Candidate
DataFormats/Common
DataFormats/JetReco
DataFormats/ParticleFlowCandidate
DataFormats/PatCandidates
DataFormats/Provenance
DataFormats/TestObjects
DataFormats/TrackReco
DataFormats/TrajectorySeed
EgammaAnalysis/PhotonIDProducers
FWCore/Framework
FWCore/Integration
HLTrigger/btau
IOPool/Common
PhysicsTools/HepMCCandAlgos
PhysicsTools/IsolationAlgos
PhysicsTools/JetMCAlgos
PhysicsTools/PatAlgos
RecoBTau/JetTagComputer
RecoEcal/EgammaClusterProducers
RecoEgamma/EgammaHFProducers
RecoEgamma/EgammaHLTProducers
RecoEgamma/EgammaPhotonProducers
RecoMET/METPUSubtraction
RecoMuon/CosmicMuonProducer
RecoMuon/GlobalMuonProducer
RecoMuon/TrackingTools
RecoTauTag/HLTProducers
RecoTauTag/ImpactParameter
RecoTauTag/RecoTau
RecoTracker/FinalTrackSelectors
RecoTracker/NuclearSeedGenerator
RecoTracker/TrackProducer
SimDataFormats/JetMatching
SimMuon/MCTruth
SimTracker/TrackAssociatorProducers
TauAnalysis/MCEmbeddingTools
Validation/RecoB


@perrotta, @cmsbuild, @civanch, @diguida, @cvuosalo, @fwyzard, @mdhildreth, @monttj, @Martin-Grunewald, @Dr15Jones, @cerminar, @deguio, @slava77, @danduggan, @vadler, @mmusich, @ktf, @nclopezo can you please review it and eventually sign? Thanks.
@ghellwig, @TaiSakuma, @jpavel, @rappoccio, @argiro, @Martin-Grunewald, @threus, @venturia, @mmarionncern, @battibass, @makortel, @acaudron, @jhgoh, @lgray, @jdolen, @ferencek, @cerati, @trocino, @rociovilar, @Sam-Harper, @wmtan, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @nhanvtran, @schoef, @mschrode, @dgulhan, @abbiendi, @imarches, @appeltel, @ahinzmann, @gpetruc, @mariadalfonso, @pvmulder, @bachtis this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@nclopezo you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

Dr15Jones commented:
Please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
Hi David, 
Could you please add a bit longer description in the subject of the PR:
- is the feature of View of Ptr complete now? 
- are there tests or examples in this PR testing or showing that?

cmsbuild commented:
-1 
Tested at: 5c6aecd6e48da3e62fea6f65cbb79350e93930d7
 I found an error when building: 
 
 <pre>>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWModelExpressionSelector.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWModelFilter.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWModelId.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWModelIdFromEveSelector.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWModelContextMenuHandler.cc: In member function 'void FWModelContextMenuHandler::showSelectedModelContext(Int_t, Int_t, FWViewContextMenuHandlerBase*) const':
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWModelContextMenuHandler.cc:269:40: error: 'kPrint' was not declared in this scope
             m_modelPopup->DisableEntry(kPrint);
                                        ^
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWNavigatorBase.cc 
gmake: *** [tmp/slc6_amd64_gcc491/src/Fireworks/Core/src/FireworksCore/FWModelContextMenuHandler.o] Error 1
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-1100/src/Fireworks/Core/src/FWNumberEntry.cc 
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3862/summary.html

Dr15Jones commented:
The test failure is unrelated to this pull request since it has shown up in other tests.

slava77 commented:
@cmsbuild please test

CMSSW_7_5_X_2015-03-26-2300 is available now and it appears to be clean from build errors
(compared to 35 errors in CMSSW_7_5_X_2015-03-26-1100)

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
a merge conflict is here now
Once that's resolved, maybe we should hold off merging other things that would interfere?

wddgit commented:
I'll look for the merge conflict and try to fix it now.

Note, this will not auto merge into ROOT5 cleanly. I will fix this after the auto merge occurs with a pull request to the ROOT5 branch.

slava77 commented:
Hi David,

Sorry, I didn't mean to suggest to have the subject of the PR so long, just the description (the first item that shows up in a larger window when a PR is submitted or the first window below the PR name)

slava77 commented:
nm, you fixed it. 
Thanks

VinInn commented on DataFormats/TrackReco/interface/TrackExtraBase.h:
```diff
@@ -51,8 +51,13 @@ class TrackExtraBase
+        TrackingRecHitRef::value_type const* item = finder(*(static_cast<TrackingRecHitRef::product_type const*>(m_hitCollection.productPtr())), m_firstHit+i);
+        return TrackingRecHitRef(m_hitCollection.id(), item, m_firstHit+i);
+      }
+      return TrackingRecHitRef(m_hitCollection,m_firstHit+i);
```
do not understand the logic here:
 what is the difference between what inside the bracket and the default return?

wddgit replied:
Ref has changed. The cached pointer it contains used to point at the container but now it points at the element in the container (much more similar to the way Ptr works and this allows Ptr to Ref conversions). The cached pointer in the RefProd continues to point at the container and m_hitCollection has a RefCore from a RefProd, so the pointer needs to be converted from pointing to the container to pointing at the element in this case.

VinInn replied:
ok for what in the bracket (a bit convoluted, but ok)
what about the default
```
+      return TrackingRecHitRef(m_hitCollection,m_firstHit+i);
```
?

wddgit replied:
The Ref contains a transient pointer as a data member. Sometimes it points at the element (in earlier versions of the code it used to point at the product), but when the Ref has just been read from an input file (and in some other circumstances) the pointer points at an EDProductGetter (a pointer which is set in a custom streamer). The EDProductGetter can get the product from the Event using the ProductID that the Ref also contains. That part of Ref has not changed.  If the function productPtr returns null, then that means the pointer in the RefCore is pointing at the EDProductGetter. And a Ref in that state can be constructed from just a RefCore and a key. In that state, RefProd and Ref contain the same pointer to the EDProductGetter. Somewhat complicated. Ask again if I did not explain it clearly enough.

wddgit committed:
Merge branch 'CMSSW_7_5_X' of https://github.com/cms-sw/cmssw into viewWithVectorOfPtr4

Resolve conflicts in:
	RecoEgamma/EgammaHLTProducers/src/HLTEcalPFClusterIsolationProducer.cc
	RecoEgamma/EgammaHLTProducers/src/HLTHcalPFClusterIsolationProducer.cc

wddgit commented:
The merge conflict should be resolved now. It was created by unrelated code merged into 7_5_X since I submitted the PR yesterday. I did not retest things but it was pretty trivial to resolve so I will leave retesting to Jenkins.

cmsbuild commented:
Pull request #8559 was updated. @perrotta, @cmsbuild, @civanch, @diguida, @cvuosalo, @fwyzard, @mdhildreth, @monttj, @Martin-Grunewald, @Dr15Jones, @cerminar, @deguio, @slava77, @danduggan, @vadler, @mmusich, @ktf, @nclopezo can you please check and sign again.

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 7939448063e982d521779910cfaa3f19f660a055
 I found an error when building: 
 
 <pre>>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATPhotonSlimmer.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATPrimaryVertexCleaner.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATSecondaryVertexSlimmer.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATSingleVertexSelector.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATPhotonProducer.cc: In member function 'virtual void pat::PATPhotonProducer::produce(edm::Event&, const edm::EventSetup&)':
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATPhotonProducer.cc:361:15: error: 'class pat::Photon' has no member named 'setEcalPFClusterIso'
       aPhoton.setEcalPFClusterIso((*ecalPFClusterIsoMapH)[photonRef]);
               ^
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/PhysicsTools/PatAlgos/plugins/PATPhotonProducer.cc:362:15: error: 'class pat::Photon' has no member named 'setHcalPFClusterIso'
       aPhoton.setHcalPFClusterIso((*hcalPFClusterIsoMapH)[photonRef]);
               ^
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3884/summary.html

cmsbuild commented:
-1 
Tested at: 5c6aecd6e48da3e62fea6f65cbb79350e93930d7
 I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 8 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/CommonTools/Utils/src/cutParser.cc 
warning: unknown warning option '-Wno-psabi' [-Wunknown-warning-option]
1 warning generated.
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/CommonTools/Utils/src/expressionParser.cc 
warning: unknown warning option '-Wno-psabi' [-Wunknown-warning-option]
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/CommonTools/Utils/src/ExpressionVar.cc:153:83: error: unused variable 'RE' [-Werror,-Wunused-variable]
  for (std::vector<edm::ObjectWithDict>::reverse_iterator RI = objects_.rbegin(), RE = objects_.rend(); RI != RI; ++RIB, ++RI) {
                                                                                  ^
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/CommonTools/Utils/src/ExpressionVarSetter.cc:1:
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/CommonTools/Utils/src/ExpressionVarSetter.h:14:
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-26-2300/src/CommonTools/Utils/src/MethodStack.h:12:
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3874/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3874/summary.html

ktf commented:
This is a genuine mistake, I guess. Should be `==` no?

slava77 commented:
@cmsbuild please test
the fix to the issue in PhysicsTools/PatAlgos is in  IB by now

cmsbuild commented:
The tests are being triggered in jenkins.

wddgit commented:
@ktf  @wmtan What is this comment referring to?

"This is a genuine mistake, I guess. Should be == no?"

I am guessing maybe it is referring to the llvm compilation
warning?  It looks like instead of RI != RI the code should be
RI != RE and it is a bug.

But this is unrelated to this PR and in the cut parser and expression
parser code. Maybe we can ignore this warning
for this PR, although someone should follow up on it and
fix it.

Also looking through this, I think the status is that we are waiting
on the tests to run again. As far as I can see this has happened
so far

1. An compilation error in PatAlgos unrelated to this PR. 
2. It was fixed and tests rerun resulting in an llvm warning
also unrelated to this PR.
3. I do not think the llvm warning was fixed, but I am not sure?
4. Tests restarted again and still running although I am not
sure why. It looks like the previous tests passed except for
the unrelated llvm warning.

Is there anything I need to do to move this forward? Did any
tests fail that I am not noticing?

slava77 commented:
it looks like jenkins is still compiling this PR, after my earlier request today.

wmtan commented:
@wddgit @ktf 
It certainly looks like "RI != RI" is wrong and should be "RI != RE".  I will investigate.  But this has nothing whatever to do with this PR, as @wddgit has said.

deguio commented:
+1
DQM only slightly touched

nclopezo commented:
Hi all,
Github is being attacked and this is causing issues with the pr tests.
On 30 Mar 2015 17:58, "deguio" <notifications@github.com> wrote:

> +1
> DQM only slightly touched
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/8559#issuecomment-87734673>.
>

cmsbuild commented:
-1 
 
Tested at: 7939448063e982d521779910cfaa3f19f660a055
I found errors in the following unit tests: 
 
 ---> test TestRunnerFWCoreTFWLiteSelector had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/2421/summary.html

wddgit committed:
Try to fix nonreproducible unit test failure

A unit test failure occurred when testing a PR
but I could not reproduce it. My guess is that
it uses a common file (test.root) name and
infrequently some unrelated tests stomps on
the file. Renamed the file and also move the
test to run in the tmp subdirectory.

wddgit commented:
I believe the unit test failure was unrelated to my changes and I could not reproduce it locally. But I tried to fix it anyway and committed the fix. Can someone try the tests again? The failing unit test was writing a file called test.root and placing it in the top level directory. I'm guessing maybe some unrelated test running in parallel was stomping on the file so I gave it a more unique name and the test is now running in the tmp subdirectory. Since I do not know how to reproduce the problem I cannot test if I fixed it.

cmsbuild commented:
Pull request #8559 was updated. @perrotta, @cmsbuild, @civanch, @diguida, @cvuosalo, @fwyzard, @mdhildreth, @monttj, @Martin-Grunewald, @Dr15Jones, @cerminar, @deguio, @slava77, @danduggan, @vadler, @mmusich, @ktf, @nclopezo can you please check and sign again.

cmsbuild commented:
-1 
Tested at: 7939448063e982d521779910cfaa3f19f660a055
 I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 8 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-29-2300/src/CommonTools/Utils/src/TMVAZipReader.cc 
warning: unknown warning option '-Wno-psabi' [-Wunknown-warning-option]
1 warning generated.
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-29-2300/src/CommonTools/Utils/src/cutParser.cc 
warning: unknown warning option '-Wno-psabi' [-Wunknown-warning-option]
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-29-2300/src/CommonTools/Utils/src/ExpressionVar.cc:153:83: error: unused variable 'RE' [-Werror,-Wunused-variable]
  for (std::vector<edm::ObjectWithDict>::reverse_iterator RI = objects_.rbegin(), RE = objects_.rend(); RI != RI; ++RIB, ++RI) {
                                                                                  ^
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-29-2300/src/CommonTools/Utils/src/ExpressionVarSetter.cc:1:
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-29-2300/src/CommonTools/Utils/src/ExpressionVarSetter.h:14:
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-03-29-2300/src/CommonTools/Utils/src/MethodStack.h:12:
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3982/summary.html

slava77 commented:
I was looking at the matrix logs and they looked like a frontier timeout:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/2421/runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECO+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECO+HARVEST+ALCATT.log
the step3 job started at 19:44 and never got past the first event after starting it at 20:43:38.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3982/summary.html

cmsbuild commented:
-1 
 
Tested at: 36f7f53e3ee2ecc4893ac37f4b05e535f5acfb90
I found errors in the following unit tests: 
 
 ---> test testDataFormatsFWLite had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/4007/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/4007/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/8559	36f7f53
code changes are in line with expectations at visual inspection.

jenkins tests of RECO workflows pass OK and show no differences in monitored quantities (fwlite or DQM based comparisons).

wddgit commented:
Thanks Slava.  Can I suggest other people go ahead and review/approve this? I was hoping for the jenkins tests to pass, but the currently running ones seem stuck and the previous 6 tries all failed for different reasons unrelated to this PR. This one was the best: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3874/summary.html. Only an unrelated llvm compiler warning and the PR to fix that was independently submitted already. (Core and DQM already approved except I put in a change to fix an unrelated unit test failure after that)

slava77 commented:
I echo David's request.
Changes are primarily technical.

davidlange6 commented:
All - i propose to merge this Saturday unless there are complaints in the meanwhile.
